src/atlas/projection/detail/StretchRegular.cc

```diff
@@ -0,0 +1,385 @@
+    /** PUT here the stretch for a point that come input
+    * give number of power as input,work it out using as from the start of regular grid.
+    * atlas::PointXY unstretchedXY = crd;
+    * atlas::PointXY stretchedXY = stretch_LAM_gen(unstretchedXY);
+    * stretch_LAM_gen(crd[]);
+    */
+
+    double lam_hires_size; ///< size regular grid x
+    double phi_hires_size; ///< size regular grid y
+    double lambda_start; ///< start grid x
+    double phi_start; ///< start grid y
+    constexpr float epsilon = std::numeric_limits<float>::epsilon(); ///< value used to check if the values are equal
+    constexpr float epstest = std::numeric_limits<float>::epsilon(); ///< correction used to change from double to integer
+
+    int n_stretchedx_ = 0, n_stretchedy_ = 0, n_x_rim_ = 0, n_y_rim_ = 0;
+
+    ///< original domain size includes the points for the rim
+    double deltax_all = (endx_ - startx_);
+    double deltay_all = (endy_ - starty_);
+
+    if (var_ratio_ == 1) {
+           lam_hires_size = deltax_all;
+           phi_hires_size = deltay_all;
+           lambda_start = x_reg_start_;
+           phi_start = y_reg_start_;
+
+         } else {
+              lam_hires_size = x_reg_end_ - x_reg_start_;
+              phi_hires_size = y_reg_end_ - y_reg_start_;
+
+              /** check for start of the new regular LAM grid x and y
+              * in the middle of the previous regular grid
+              */
+
+              ///< distance end of the grid and internal regular grid
+              double add_xf_ = (deltax_all + epstest - lam_hires_size) / 2.;
+              double add_yf_ = (deltay_all + epstest - phi_hires_size) / 2.;
+              /**
+               *  Compute the number of points for different part of the grid
+               *  internal regular grid high resolution
+               *  stretched grid
+               *  external regular grid low resolution
+               *  +1 otherwise I have the intervals not the points
+              */
+              n_x_rim_ = rim_widthx_/delta_low_ ;
+              n_y_rim_ = rim_widthy_/delta_low_ ;
+              n_stretchedx_ = ((deltax_all + epstest - lam_hires_size) / delta_high_) - n_x_rim_;
+              n_stretchedy_ = ((deltay_all + epstest - phi_hires_size) / delta_high_) - n_y_rim_ ;
+              lambda_start = x_reg_start_;
+              phi_start = y_reg_start_;
+              /**
+               *  check if stretched grid is in the middle
+               *  of the previous regular grid
+              */
+              double check_x = startx_ + add_xf_ - lambda_start;
+              double check_y = starty_ + add_yf_ - phi_start;
+              double check_st = n_stretchedx_ - n_stretchedy_;
+              checkvalue(epsilon, check_x);
+              checkvalue(epsilon, check_y);
+              checkvalue(epsilon, check_st);
+
+         }
+
+    int nx_ = ((deltax_all + epstest) / delta_high_) + 1;
+    int ny_ = ((deltay_all + epstest) / delta_high_ ) + 1;
```

Comment:
The selection is identical for each point that will be projected.
It will be more performant to add this into the constructor instead.

Reply:
Hi Willelm,
Thank you for your comments.
The code is still not pushed. This is a general question.
I put the code under the header for the function xy2lonlat. The issue is that the function itself is defined as virtual const, I set the variables I need as mutable. It works but I have a second problem.
Hi all, I changed the header of a file in atlas from:

in the header I have
void xy2lonlat( double crd[]) const override{
mycode
};

In the cpp file I have:
void StretchLAM<Rotation>::xy2lonlat( double crd[] ) const {
some other code
};

the problem is that now I have the error:
"error: redefinition of 'void atlas::projection::detail::StretchLAM<Rotation>::xy2lonlat(double*) const'"

Do you have any idea how to solve this?

Has somebody an idea how to avoid it?

Reply:
In the header you only need the signature of the function:
```c++
void xy2lonlat(double crd[]) const override;
```
In the .cc file comes the implementation with all the code:
```c++
void StretchLAM::xy2lonlat(double crd[]) const {
   <the implementation>
}
The function is correctly marked `const` and using mutable member variables is a hack, which is dangerous in e.g. an OpenMP context.
You have to initialize these values in the constructor instead, which I think you can because they are independent of the `crd` argument.

Reply:
I implemented the code in the wrong construnctor. Sorry.
Now is in the file.cc for StretchLAM

Reply:
I also didn't need the use of mutable member, I pushed now my changes.
Thanks